tests/4.5/taskloop/test_taskloop_final.c

```diff
@@ -0,0 +1,107 @@
+//--------------------------------------------- test_taskloop_final.c ----------------------------------------//
+//
+// OpenMP API Version 4.5 August 2015
+//
+// This test checks the taskloop directive with the if clause specified.
+// The 'taskloop' construct parallelize loops with independent iterations by creating tasks. 
+// It allows for efficient parallel execution of loop iterations by distributing them among multiple threads. 
+// The 'final clause is used to ensure that all tasks in taskloop are run by same thread if the 
+// final(expression) is TRUE, else different threads could execute the tasks.      
+//----------------------------------------------------------------------------------------------------------//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define NUM_TASKS 10
+#define NUM_THREADS 10
+
+#define M 100
+
+int thread_ids[NUM_TASKS][3]; //Array 2nd diemention value i.e 3 is depend on number of omp task under taskloop
+
+
+//if all thread ids are same, val=1 else val=0
+int thread_ids_are_same_in_task() {
+
+ 	int val = -1; 
+ 	for(int i = 0; i < NUM_TASKS; i++) 
+	{
+	 	int first = thread_ids[i][0];
+
+	 	for(int j = 1; j < 3; j++) 
+	 	{
+	        	if(first != thread_ids[i][j]) {
+		  		val = 0;  //if all the thread ids are not same per task
+		  		break; 
+			}
+			else if(first == thread_ids[i][j]) {
+		    		val = 1; //if all the thread ids are same per task
+			}
+   		}
+	 
+		if(val == 0)
+	       		break;	
+   	}
+
+	return val;
+}
+
+int test_taskloop_final(int THRESHOLD) {
+  
+  int errors = 0; 
+  
+  #pragma omp parallel num_threads(NUM_THREADS)
+  {
+  	#pragma omp single
+    	{	
+        	#pragma omp taskloop final(M == THRESHOLD)
+      		for (int i = 0; i < NUM_TASKS; i++) {
+			#pragma omp task 
+			{
+	 			thread_ids[i][0] = omp_get_thread_num();
+			}
+			#pragma omp task 
+			{
+	 			thread_ids[i][1] = omp_get_thread_num();
+			}
+
+			#pragma omp task 
+			{
+	 			thread_ids[i][2] = omp_get_thread_num();
+			}
+      		} 
+   	}
+  }
+
+ int val = thread_ids_are_same_in_task();
+ 
+  if (M == THRESHOLD) {
+    if (val != 1) {
+      errors++; 
+    }
+  } else if (M != THRESHOLD) {
+    if (val != 0) {
```

Comment:
Is it still allowed that the same single thread executes all the generated tasks even if the final clause expression evaluates to false?

Reply:
'if' the 'final' expression is FALSE, its not sure that same thread runs a group of tasks. It is sure that taskloop tasks are not properly grouped.

Reply:
It is still legal for an implementation to use a single thread to execute all the generated tasks, and thus we cannot say `val != 0` is an error here (the `final` expression is false).

Reply:
Even if multiple threads are used to execute the `taskloops`, it is still possible to get `val !=0` when the `final` expression is false.

Reply:
@shivapdr , the point @seyonglee  is making is that when FINAL is false, the ERROR should change to a WARNING *if* the tasks were executed by the same thread as it is still ;legal to do so.

Reply:
d91f87c27f75ada3bdb59da92bb5fb220f901238